rplugin/python3/deoplete/sources/deoplete_jedi.py

```diff
@@ -47,21 +57,108 @@ def get_complete_position(self, context):
         line = context['position'][1]
         col = context['complete_position']
         buf = self.vim.current.buffer
-        source = '\n'.join(buf[:])
+        source = buf[:]
         cline = self.vim.current.line
 
+        extra_modules = []
+        cache_key = None
+        deoplete_input = context['input'].strip()
+        cache_line = 0
+
+        if self.cache_enabled:
+            # Caching based on context input.  If the input is blank, it was
+            # triggered with `.` to get module completions.
+            #
+            # The module files as reported by Jedi are stored with their
+            # modification times to help detect if a cache needs to be
+            # refreshed.
+            #
+            # For scoped variables in the buffer, construct a cache key using
+            # the filename.  The buffer file's modification time is checked to
+            # see if the completion needs to be refreshed.  The approximate
+            # scope lines are cached to help invalidate the cache based on line
+            # position.
+
+            if context.get('complete_str'):
+                # Note: Module completions will be an empty string.
+                word = context.get('complete_str')
+                cache_key = buf.name
+                extra_modules.append(buf.name)
+                cache_line = line - 1
+            else:
+                dot = deoplete_input.rfind('.')
+                if dot != -1:
+                    # Only cache module results.
+                    cache_key = deoplete_input[:dot]
+                    if cache_key.startswith('self'):
+                        # TODO: Get class lines and cache these differently
+                        # based on cursor position.
+                        # Cache `self.`, but monitor buffer file's modification
+                        # time.
+                        extra_modules.append(buf.name)
+                        cache_key = '{}.{}'.format(buf.name, cache_key)
+                        cache_line = line - 1
+                else:
+                    extra_modules.append(buf.name)
+                    cache_key = '{}.{}'.format(buf.name, deoplete_input)
+
+            if cache_key and cache_key in self.cache:
+                # XXX: Hash cache keys to reduce length?
+                cached = self.cache.get(cache_key)
+                lines = cached.get('lines', [0, 0])
+                modules = cached.get('modules')
+                # TODO: If cache is invalid, return stale results and start
+                # a thread to refresh.
+                if cache_line >= lines[0] and cache_line <= lines[1] \
+                        and all([filename in modules for filename in
+                                 extra_modules]) \
+                        and all([int(os.path.getmtime(filename)) == mtime
+                                 for filename, mtime in modules.items()]):
+                    self.debug('cache hit')
```

Comment:
I forgot I left those in.  I'll remove.

> WDYT?

I think since deoplete.nvim is driven with Python, it should use python's logging module to log to a file.  Checking Vim's message log can be frustrating.